File: helo/src/main/java/cloudcode/bfn/web/FirebaseScaffold.java
package cloudcode.bfn.web;

import cloudcode.bfn.BFNWebApplication;
import com.google.api.core.ApiFuture;
import com.google.cloud.firestore.Firestore;
import com.google.cloud.firestore.QueryDocumentSnapshot;
import com.google.cloud.firestore.QuerySnapshot;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.cloud.FirestoreClient;
import com.google.firebase.messaging.FirebaseMessaging;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.slf4j.Logger;
import org.springframework.stereotype.Component;
import com.google.auth.oauth2.GoogleCredentials;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.concurrent.ExecutionException;
import javax.annotation.PostConstruct;
@Component
public class FirebaseScaffold {

    private final static Logger logger = LoggerFactory.getLogger(FirebaseScaffold.class.getSimpleName());
    private Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    @Autowired
    private ApplicationContext context;
    @PostConstruct
    public void init() throws IOException {
        logger.info("\n\n\uD83D\uDC13 \uD83D\uDC13 \uD83D\uDC13 FirebaseScaffold  \uD83C\uDF51 " +
                "init: starting to fuck with \uD83D\uDC13  \uD83C\uDF51 Firebase  \uD83C\uDF51 \uD83D\uDC13 \uD83D\uDC13 \n");
        logger.info("\uD83D\uDC7D \uD83D\uDC7D \uD83D\uDC7D \uD83D\uDC7D " +
                "PostConstruct: \uD83C\uDF3F Alexa and AI are coming for you!  \uD83C\uDF4F "
                        .contains(context.getApplicationName().concat("  \uD83C\uDF4F ")));

        logger.info("\uD83D\uDC4C\uD83C\uDFFE \uD83D\uDC4C\uD83C\uDFFE YEBO! appProperties are cool. " +
                "\uD83C\uDF4F \uD83C\uDF4F \uD83C\uDF4F this shit is hanging in there ..." );

        logger.info("\uD83D\uDD06 \uD83D\uDD06 \uD83D\uDD06  FirebaseScaffold: setting up Firebase service account ...."
                + " \uD83D\uDD06 \uD83D\uDD06 \uD83D\uDD06");
        try {
            FirebaseOptions options = new FirebaseOptions.Builder()
                    .setCredentials(GoogleCredentials.getApplicationDefault())
                    .setProjectId("bfn-mobile-backend")
                    .setDatabaseUrl("https://bfn-mobile-backend.firebaseio.com").build();
            FirebaseApp app = FirebaseApp.initializeApp(options);
            logger.info("\uD83E\uDDE9\uD83E\uDDE9\uD83E\uDDE9  \uD83E\uDDE9\uD83E\uDDE9\uD83E\uDDE9 "
                    + "Firebase Admin SDK Setup OK:  \uD83E\uDDE9\uD83E\uDDE9\uD83E\uDDE9 app: "
                    + app.toString());
            //
            //listAccountsFromFirebase();
            logger.info("\uD83D\uDD06  \uD83D\uDD06  \uD83D\uDD06  Getting bean to start refreshing nodes on Firestore  \uD83D\uDD06 ");
            try {
                BFNController adminController = context.getBean(BFNController.class);
                logger.info(("\uD83C\uDF4E Forcing BFN to say Hello!  \uD83D\uDC38 \uD83D\uDC38 \uD83D\uDD90\uD83C\uDFFD " +
                        "\uD83D\uDD90\uD83C\uDFFD \uD83D\uDD90\uD83C\uDFFD ")
                        .concat(adminController.helloWorld()));
            } catch (Exception e) {
                logger.error(e.getMessage());
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: bfn-kotlin-dlt/clients/src/main/kotlin/com/template/webserver/FirebaseScaffold.kt
package com.template.webserver

import com.google.auth.oauth2.GoogleCredentials
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.io.FileInputStream
import javax.annotation.PostConstruct

@Component
class FirebaseScaffold {
    private val logger = LoggerFactory.getLogger(FirebaseScaffold::class.java)
    @Autowired
    private lateinit var appProperties: AppProperties

    @PostConstruct
    fun init() {
        logger.info("\uD83D\uDC7D \uD83D\uDC7D \uD83D\uDC7D \uD83D\uDC7D " +
                "PostConstruct: \uD83C\uDF3F Alexa and AI are coming for you!")
        var path: String = "/Users/aubs/WORK/CORDA/bfn.json";
        if (appProperties == null) {
            logger.info("WTF! \uD83D\uDC7A \uD83D\uDC7A this shit is still null \uD83D\uDC7A ")
            throw java.lang.Exception("Houston, we are dead!")
        } else {
            logger.info("\uD83D\uDC4C\uD83C\uDFFE \uD83D\uDC4C\uD83C\uDFFE YEBO! " +
                    "\uD83C\uDF4F \uD83C\uDF4F \uD83C\uDF4F this shit is hanging in there, " +
                    "firebasePath: ${appProperties.firebasePath}")
            path = appProperties.firebasePath
        }
        logger.info("\uD83D\uDD06  \uD83D\uDD06  \uD83D\uDD06  BFNWebApi:  setting up Firebase service account ...."
                + " \uD83D\uDD06  \uD83D\uDD06  \uD83D\uDD06")
        try {
            logger.info("\uD83D\uDD06  \uD83D\uDD06  \uD83D\uDD06 "
                    + " PATH for Firebase Service Account: \uD83D\uDC99  "
                    + path)
            val serviceAccount = FileInputStream(path)
            val options = FirebaseOptions.Builder()
                    .setCredentials(GoogleCredentials.fromStream(serviceAccount))
                    .setDatabaseUrl("https://bfn-mobile-backend.firebaseio.com").build()
            val app = FirebaseApp.initializeApp(options)
            logger.info(" \uD83E\uDDE9\uD83E\uDDE9\uD83E\uDDE9  \uD83E\uDDE9\uD83E\uDDE9\uD83E\uDDE9 "
                    + "Firebase Admin SDK Setup OK:  \uD83E\uDDE9\uD83E\uDDE9\uD83E\uDDE9 name: "
                    + app.name)
            listAccountsFromFirebase()
        } catch (e: Exception) {
            logger.error(" \uD83D\uDC7F  \uD83D\uDC7F  \uD83D\uDC7F  \uD83D\uDC7F Firebase Admin SDK setup failed")
            throw Exception(" \uD83D\uDC7F  \uD83D\uDC7F unable to set Firebase up", e)
        }

    }
    private fun listAccountsFromFirebase() {
        val users = FirebaseUtil.users
        logger.info("\n\n⚽️ Accounts on Firebase Auth:  \uD83D\uDC9A \uD83D\uDC99 ${users.size}")
        var cnt = 0;
        val sorted = users.sortedBy { it.displayName }
        sorted.forEach() {
            cnt++
            logger.info("⚽️ Account: #$cnt \uD83D\uDC9A \uD83D\uDC99 ${it.displayName} - \uD83D\uDD06 ${it.email}")
        }
    }
}
